src/controller/stream.rs

```diff
@@ -0,0 +1,341 @@
+                        "Got empty response".to_string(),
+                    ));
+                } else {
+                    tracing::debug!(
+                        "Got response for chunk {} with blocks {}-{}. {}/{} blocks fetched from this chunk. Current response length: {}",
+                        index,
+                        current_range.begin,
+                        last_block,
+                        last_block - full_range.begin as u64 + 1,
+                        full_range.end - full_range.begin + 1,
+                        accum_response.len()
+                    );
+                    current_range.begin = (last_block + 1) as u32;
+                }
+            }
+        }
+    }
+
+    fn get_one_response(
+        request: ClientRequest,
+        network: Arc<NetworkClient>,
+        timeouts: Arc<TimeoutManager>,
+        index: usize,
+        range: Range,
+    ) -> impl Future<Output = Result<query_result::Result, RequestError>> {
+        // TODO: don't modify the query
+        let send_query = move || {
+            let start_time = tokio::time::Instant::now();
+            Self::query_some_worker(
+                network.clone(),
+                request.query.with_range(&range),
+                request.dataset_id.clone(),
+                index,
+                range,
+            )
+            .map(move |result| (result, start_time))
+        };
+        async move {
+            let timeouts = timeouts.clone();
+            let mut requests = JoinSet::new();
+            let mut tries_left = request.retries;
+            for _ in 0..request.request_multiplier {
+                requests.spawn(send_query());
+                tries_left -= 1;
+            }
+            loop {
+                let (send_result, start_time) = tokio::select! {
+                    _ = timeouts.sleep(), if tries_left > 0 => {
+                        tracing::debug!("Request for chunk {index} is running for too long, sending one more query");
+                        requests.spawn(send_query());
+                        tries_left -= 1;
+                        continue;
+                    }
+                    result = requests.join_next() => {
+                        result
+                            .expect("No downloads left for chunk")
+                            .expect("Awaiting download panicked")
+                    }
+                };
+                let result = match send_result {
+                    Ok(result) => result,
+                    Err(e) => {
+                        if tries_left == 0 {
+                            return Err(e);
+                        }
+                        tries_left -= 1;
+                        tracing::debug!(
+                            "Couldn't schedule query ({e:?}), retrying in {:?}",
+                            request.backoff
+                        );
+                        let backoff = request.backoff;
```

Comment:
It's also useful when all the workers are unavailable for requests at the moment. Without a backoff you would just see "couldn't find worker for chunk {}" repeatedly and the stream would end because of a 404 error.

Reply:
It may be a bit messy in the code, but there are two kinds of errors here:
- Errors returned immediately from `query_some_worker` function. I call those "query scheduling errors". They occur immediately, without `await`ing any async tasks. Therefore, retrying them immediately will probably cause the same result, so I've introduced a backoff.
- Errors received from the network, after `await`ing something. Those should probably be retried immediately after we get the error.

Reply:
>So I think the usage of backoff doesn't really depend on which worker is queried.

That's not true. 

What is the most common cause of query failure according to our expectations? It should be a problem with a single worker or connectivity with a single worker. In such case a backoff timeout is just a waste of time.

>transport message queue is full, "query scheduling errors"

Sounds like something, that should not happen in the first place. But, if something like this happens, I'd say the best way to deal with that is to return the error immediately without any retries.

Reply:
> In such case a backoff timeout is just a waste of time
Okay, agreed. But then a better approach would be to treat "not possible to send query" and "query failed" errors separately. I'll try to find a better solution. I don't really like how this function is implemented at the moment.

>Sounds like something, that should not happen in the first place
It can easily happen because we're scheduling multiple requests without an `await` point in between. So, for example, on a single-threaded runtime with buffer size > msg_queue_size it's guaranteed to happen. You're right, it's better to return such error and deal with it on a higher level.

Reply:
I've handled the "not possible to send query" type of error separately. Now the buffer won't grow until the new queries can be successfully scheduled. It helps to avoid interrupting the stream if the `buffer_size` is too large for the current network conditions (many overloaded workers).

However, the transport queue (1000 messages) is filled up quite quickly and it probably becomes the bottleneck if multiple streams are run in parallel :(
@wiezzel is it safe to increase the outgoing messages buffer size?